agents/ahv/fence_ahv.py

```diff
@@ -0,0 +1,608 @@
+            logging.error("Failed to power on VM %s", vm_uuid)
+            raise AHVFenceAgentException from err
+        except AHVFenceAgentException as err:
+            logging.error("Failed to retrieve etag from headers")
+            raise AHVFenceAgentException from err
+
+        return resp
+
+    def _wait_for_task(self, task_uuid, timeout=None):
+        if not task_uuid:
+            logging.error("Task UUID was not provided")
+            raise AHVFenceAgentException("Task UUID not provided")
+
+        task_url = f"{self.task_url}/{task_uuid}"
+        header_str = self._get_headers()
+        task_resp = None
+        interval = 5
+        task_status = None
+
+        if not timeout:
+            timeout = MIN_TIMEOUT
+        else:
+            try:
+                timeout = int(timeout)
+            except ValueError as err:
+                timeout = MIN_TIMEOUT
+
+        while task_status != 'SUCCEEDED':
+            if timeout <= 0:
+                raise TaskTimedOutException(f"Task timed out: {task_uuid}")
+
+            time.sleep(interval)
+            timeout = timeout - interval
+
+            try:
+                task_resp = self.request(url=task_url, method='GET',
+                                         headers=header_str, verify=self.verify)
+                task_status = task_resp.json()['data']['status']
+            except NutanixClientException as err:
+                logging.error("Unable to retrieve task status")
+                raise AHVFenceAgentException from err
+            except Exception as err:
+                logging.error("Unknown error")
+                raise AHVFenceAgentException from err
+
+            if task_status == 'FAILED':
+                raise AHVFenceException(f"Task failed, task uuid: {task_uuid}")
+
+    def list_vms(self, filter_str=None, limit=None):
+        vms = None
+
+        try:
+            vms = self._get_all_vms(filter_str, limit)
+        except NutanixClientException as err:
+            logging.error("Failed to retrieve VM list")
+            raise AHVFenceAgentException from err
+
+        if not vms or not isinstance(vms, dict):
+            logging.error("Failed to retrieve VM list")
+            raise AHVFenceAgentException from err
+
+        if 'data' not in vms:
+            err = "Error: No VMs listed in response"
+            logging.error("Failed to retrieve VM list")
+            raise AHVFenceAgentException(err)
+
+        vm_list = {}
+
+        for vm in vms['data']:
+            vm_name = vm['name']
+            ext_id = vm['extId']
+            power_state = vm['powerState']
+            vm_list[vm_name] = (ext_id, power_state)
+
+        return vm_list
+
+    def get_power_state(self, vm_name=None, vm_uuid=None):
+        resp = None
+        power_state = None
+
+        if not vm_name and not vm_uuid:
+            logging.error("Require at least one of VM name or VM UUID")
+            raise AHVFenceAgentException("No arguments provided")
```

Comment:
You should probabl fail with e.g. EC_BAD_ARGS here.

Reply:
Done.